main.cpp

```diff
@@ -322,11 +390,61 @@ int uthread_terminate(int tid) {
     } else if((it = find(waiting_list.begin(), waiting_list.end(), tid)) != waiting_list.end()) {
 	waiting_list.erase(it);
     } else {
-	return -1;
+	enable_interrupts();
+	thread_complete();  // If tid is the running thread
+	return 0;
     }
 
     free_waiting_threads(tid);
+
+    enable_interrupts();
+    return 0;
+}
+
+/**
+ * Sets the time slice for how long each thread runs
+ * @param time_slice - the new time slice for each thread in microseconds 
+ */
+int uthread_init(int time_slice) {
+    timer.it_interval.tv_usec = time_slice;
+    return 0;
+}
+
+/**
+ * Calls thread_yield every time the timer expires
+ */
+void timer_handler(int signum) {
+    static int count = 0;
+    printf("Timer expired: %d\n", ++count);
+    uthread_yield();
+}
+
+/**
+ * Set up timer
+ */
+int setupitimer(void) {
+    timer.it_interval.tv_sec = 0;
+    timer.it_interval.tv_usec = 1000;
+    timer.it_value = timer.it_interval;
+    return setitimer(ITIMER_VIRTUAL, &timer, nullptr);
```

Comment:
Any ideas on what the initial value for the timer interval should be?

Reply:
1 sec looks good